src/raftstore/store/worker/region.rs

```diff
@@ -266,13 +287,139 @@ impl SnapContext {
                 escape(&end_key),
                 e
             );
+            RANGE_DELETION_COUNTER_VEC
+                .with_label_values(&["fail"])
+                .inc();
+        } else {
+            info!(
+                "succeed in deleting data in [{}, {})",
+                escape(&start_key),
+                escape(&end_key),
+            );
+            RANGE_DELETION_COUNTER_VEC
+                .with_label_values(&["success"])
+                .inc();
+        }
+    }
+}
+
+#[derive(Default)]
+struct PendingDeleteRanges {
+    ranges: BTreeMap<Vec<u8>, (Vec<u8>, time::Instant)>, // range start_key, (range end_key, timeout)
+}
+
+impl PendingDeleteRanges {
+    // find out the ranges that need to be trimed or discarded
+    fn find_overlap_ranges(&self, start_key: Vec<u8>, end_key: Vec<u8>) -> Vec<(Vec<u8>, Vec<u8>)> {
+        let mut ranges = Vec::new();
+        // find the first range that may overlap with [start_key, end_key)
+        let sub_range = self.ranges.range((Unbounded, Excluded(start_key.clone())));
+        if let Some((s_key, &(ref e_key, _))) = sub_range.last() {
+            if e_key > &start_key {
+                ranges.push((s_key.clone(), e_key.clone()));
+            }
+        }
+
+        // find the rest ranges that overlap with [start_key, end_key)
+        for (s_key, &(ref e_key, _)) in self.ranges.range((Included(start_key), Excluded(end_key))) {
+            ranges.push((s_key.clone(), e_key.clone()));
         }
+        ranges
+    }
+
+    // filter out the overlap in ranges, after this function the ranges will not overlap
+    pub fn remove(&mut self, start_key: Vec<u8>, end_key: Vec<u8>) {
+        if start_key >= end_key {
+            return;
+        }
+        RANGE_DELETION_COUNTER_VEC
```

Comment:
It's better to move these metrics out of `PendingDeleteRanges`, `PendingDeleteRanges` is a lib.